cpp/src/decisiontree/decisiontree.cu

```diff
@@ -14,741 +14,295 @@
-  }
-}
-
-//Class specializations
-template class DecisionTreeBase<float, int>;
-template class DecisionTreeBase<float, float>;
-template class DecisionTreeBase<double, int>;
-template class DecisionTreeBase<double, double>;
-
-template class DecisionTreeClassifier<float>;
-template class DecisionTreeClassifier<double>;
-
-template class DecisionTreeRegressor<float>;
-template class DecisionTreeRegressor<double>;
-
-}  //End namespace DecisionTree
-
-// Stateless API functions
-
-// ----------------------------- Classification ----------------------------------- //
-
-/**
- * @brief Build (i.e., fit, train) Decision Tree classifier for input data.
- * @param[in] handle: cumlHandle
- * @param[in,out] dt_classifier: Pointer to Decision Tree Classifier object. The object holds the trained tree.
- * @param[in] data: train data in float (nrows samples, ncols features) in column major format, excluding labels. Device pointer.
- * @param[in] ncols: number of features (i.e., columns) excluding target feature.
- * @param[in] nrows: number of training data samples of the whole unsampled dataset.
- * @param[in] labels: 1D array of target features (int only). One label per training sample. Device pointer.
-				  Assumption: labels need to be preprocessed to map to ascending numbers from 0;
-				  needed for current gini impl. in decision tree
- * @param[in,out] rowids: This array consists of integers from (0 - n_sampled_rows), the same array is then rearranged when splits are made. This allows, us to contruct trees without rearranging the actual dataset. Device pointer.
- * @param[in] n_sampled_rows: number of training samples, after sampling. If using decsion tree directly over the whole dataset (n_sampled_rows = nrows)
- * @param[in] n_unique_labels: #unique label values. Number of categories of classification.
- * @param[in] tree_params: Decision Tree training hyper parameter struct
- */
-void fit(const ML::cumlHandle &handle,
-         DecisionTree::DecisionTreeClassifier<float> *dt_classifier,
-         float *data, const int ncols, const int nrows, int *labels,
-         unsigned int *rowids, const int n_sampled_rows, int unique_labels,
-         DecisionTree::DecisionTreeParams tree_params) {
+void decisionTreeClassifierFit(const ML::cumlHandle &handle,
+                               TreeClassifierF *&tree, float *data,
+                               const int ncols, const int nrows, int *labels,
+                               unsigned int *rowids, const int n_sampled_rows,
+                               int unique_labels,
+                               DecisionTree::DecisionTreeParams tree_params) {
+  std::shared_ptr<DecisionTreeClassifier<float>> dt_classifier =
+    std::make_shared<DecisionTreeClassifier<float>>();
   dt_classifier->fit(handle, data, ncols, nrows, labels, rowids, n_sampled_rows,
-                     unique_labels, tree_params);
+                     unique_labels, tree, tree_params);
+  dt_classifier.reset();
```

Comment:
@myrtw the reset automatically happens once this function finishes. So, better to remove it.

Reply:
Same for all others too.

Reply:
Sure, will remove it for all. I was being too explicit.